domain/model/service/service.go

```diff
@@ -542,3 +589,53 @@ func (s *Service) UpdateCredential(
 
 	return s.st.UpdateCredential(ctx, uuid, key)
 }
+
+func getWatchActivatedModelsMapper(st State, logger logger.Logger) func(ctx context.Context, db database.TxnRunner,
+	changes []changestream.ChangeEvent) ([]changestream.ChangeEvent, error) {
+
+	return func(ctx context.Context, db database.TxnRunner,
+		changes []changestream.ChangeEvent) ([]changestream.ChangeEvent, error) {
+		uuidToChangeEventMap := make(map[string]changestream.ChangeEvent)
+		modelUUIDs := make([]string, len(changes))
+
+		for i, change := range changes {
+			modelUUID := change.Changed()
+			modelUUIDs[i] = modelUUID
+			uuidToChangeEventMap[modelUUID] = change
+		}
+
+		// Retrieve all activate status of all models with associated uuids
+		activatedModelUUIDs, err := st.GetActivatedModelUUIDs(ctx, modelUUIDs)
+
+		// There should be no errors returned if there are no activated models found.
+		if err != nil {
+			logger.Errorf(ctx, "failed to retrieve activated model UUIDs for UUIDs %v: %v", modelUUIDs, err)
+			return nil, err
+		}
+
+		activatedModelChangeEvents := make([]changestream.ChangeEvent, 0, len(activatedModelUUIDs))
+		// Add all events associated with activated model UUIDs
+		for _, activatedModelUUID := range activatedModelUUIDs {
+			if changeEvent, exists := uuidToChangeEventMap[activatedModelUUID.String()]; exists {
+				activatedModelChangeEvents = append(activatedModelChangeEvents, changeEvent)
+			}
+		}
+
+		return activatedModelChangeEvents, nil
```

Comment:
This isn't correct, or at least this doesn't have the desired consequences that the original changes has.

If two models are created at different times but are activated at the same time (this is a potential scenario), then this will not guarantee the original order of the changes.

It might not matter, but downstream it might. We should always preserve change stream event ordering.

Reply:
I see, thanks for pointing this out Simon. Previously, @tlm and I discussed the possibility of exploring batch querying. Would it be best to wait for his input before proceeding? I can check with him on this tomorrow. Happy to adjust the implementation to whichever approach makes the most sense. :)

Discussion link:
https://github.com/juju/juju/pull/19071#discussion_r1982402542

Reply:
@tlm is right, you don't need a map here, just remove events from the slice that aren't activated. That way you keep the order.

Reply:
Oh I see, I got what you mean now. Sorry might have overthought this. I have edited the code accordingly. But I think we could still use a map for the retrieved activated model UUIDs? For looking up if the change event is activated when iterating changes. I have adjusted in the latest commit, please let me know if it makes sense and I'd be happy to address if there's any issues, thanks alot Simon :D